---
title: "R Notebook Template"
---
```{r}
##########   Packages and Settings   ##########
#The R packages I want to load
packages = c('BiocManager',
						'rmarkdown', 'svglite', 
						'extrafont',
						'rotl', 'ape', 'phytools',
            'jsonlite', 'readxl',
						'eList',
            'stringr',
            'dplyr', 'tidyverse',
            'corrplot', 'cluster', 'factoextra',
            'ggplot2', 'ggraph', 'scales',
						'gt', 'gtExtras',
						'phytools'
			)

#Require each package, but before that, check if it's installed. If not, install it first.
for (pkg in packages) {
  if (!require(pkg, character.only=TRUE, quietly=TRUE)){
    if(pkg=='DESeq2' | pkg=='tximport'){
      BiocManager::install(pkg)
    }
    else{
      install.packages(pkg, character.ony=TRUE)
    }
  }

  library(pkg, character.only = TRUE)
}
loadfonts(quiet=TRUE)
```



```{r}
##########   Initial Data Load   ##########
orthos = read_xlsx("Findings.xlsx", range="A1:K100", col_types = c("text", "text", "skip", "text", "text", "numeric", "skip", "skip", "skip", "skip", "numeric")) %>% filter(!if_all(.fns=is.na))
```



```{r}
ggplot(orthos, aes(x=Subunit_count)) +
	geom_histogram(bins = 50, color='black', fill='darkgrey') +
	scale_x_continuous(limits=c(0,50), breaks = seq(0,50,5), oob=squish) +
	scale_y_continuous(limits=c(0,15))  +
	labs(x="Ubiquitin subunits",
			 y="Number of genes") +
	theme(text=element_text(family='Arial'))
# ggsave('visuals/Orthos_subunit_hist.png', dpi=500)
ggsave('visuals/Orthos_subunit_hist.svg')
```



```{r}
tol_response = tnrs_match_names(orthos$Species_sci_name)

taxon_info = taxonomy_taxon_info(tol_response$ott_id)
for(taxon in taxon_info){
	stopifnot(taxon$rank %in% c("species", "subspecies"))
	stopifnot(taxon$is_suppressed == FALSE)
}
```



```{r}
suppressWarnings(tree <- tol_induced_subtree(ott_ids=tol_response$ott_id, label_format = "name"))
str(tree)
```



```{r}
tree$tip.label = gsub("_", " ", tree$tip.label)
regex = "\\s\\([^)]*\\)"
gsub(regex, "", "Capra hircus (species in domain Eukaryota)") == "Capra hircus"
tree$tip.label = gsub(regex, "", tree$tip.label)
str(tree)
tree$tip.label[6] = "Pan troglodytes"
tree$tip.label[21] = "Danio Rerio"
tree$tip.label[24] = "Astatotilapia calliptera"

for(i in 1:length(tree$tip.label)){
	orthos[orthos$Species_sci_name == tree$tip.label[i], "Tree_order"] = length(tree$tip.label) - i + 1
	
	common_name = orthos[orthos$Species_sci_name == tree$tip.label[i], "Species_common_name"]
	ubq_count = orthos[orthos$Species_sci_name == tree$tip.label[i], "Subunit_count"]
	spaces = orthos[orthos$Species_sci_name == tree$tip.label[i], "Tree_spaces"]

	if(i==24){
		common_name = "Eastern Happy*"
		ubq_count = "9,46"
		spaces = 57
	}
	
	padding_spaces = paste(rep(" ", spaces), collapse="")
	tree$tip.label[i] = paste(common_name, padding_spaces, ubq_count, sep="")
}
str(tree)
```



```{r}
# png("visuals/ubc_ortho_tree.png", width=4800, height=4800)
pdf("visuals/ubc_ortho_tree.pdf", width=64, height=64)
plot.phylo(tree, font=1, edge.width = 4, cex=6, family="mono", x.lim=300)
text(x=159, y=46.2, cex=7, bquote(underline("Ubiquitin subunits")))
text(x=20, y=0, cex=4, substitute(paste("* The Eastern Happy has 2 ", italic("UBC"), " orthologs")))
dev.off()
```
